In [1]:
import matplotlib.pyplot as plt
from matplotlib.pyplot import imshow
from IPython.display import Image, display

import pandas as pd
import numpy as np
import json
%matplotlib notebook

# Load Json File

In [2]:
f = open('/projects/kawa7978/layer1.json', 'r')
layer1 = json.load(f)

In [3]:
f = open('/projects/kawa7978/layer2.json', 'r')
layer2 = json.load(f)

In [4]:
f = open('/projects/kawa7978/det_ingrs.json', 'r')
ingr = json.load(f)

# Layer1
## Structure
Layer1 = List of layers [layer, layer, ..., layer]

Layer
```js
{
  id: String,  // unique 10-digit hex string
  title: String,
  instructions: [ { text: String } ],
  ingredients: [ { text: String } ],
  partition: ('train'|'test'|'val'),
  url: String
}
```

In [5]:
print('No. of Layers: %i'%(len(layer1)))
print('-'*100)
print('Example: index==0')
print('-'*100)

for k, v in layer1[0].items():
    print(k)
    print(v)
    print('-'*20)
print('-'*100)


No. of Layers: 1029720
----------------------------------------------------------------------------------------------------
Example: index==0
----------------------------------------------------------------------------------------------------
ingredients
[{'text': '6 ounces penne'}, {'text': '2 cups Beechers Flagship Cheese Sauce (recipe follows)'}, {'text': '1 ounce Cheddar, grated (1/4 cup)'}, {'text': '1 ounce Gruyere cheese, grated (1/4 cup)'}, {'text': '1/4 to 1/2 teaspoon chipotle chili powder (see Note)'}, {'text': '1/4 cup (1/2 stick) unsalted butter'}, {'text': '1/3 cup all-purpose flour'}, {'text': '3 cups milk'}, {'text': '14 ounces semihard cheese (page 23), grated (about 3 1/2 cups)'}, {'text': '2 ounces semisoft cheese (page 23), grated (1/2 cup)'}, {'text': '1/2 teaspoon kosher salt'}, {'text': '1/4 to 1/2 teaspoon chipotle chili powder'}, {'text': '1/8 teaspoon garlic powder'}, {'text': '(makes about 4 cups)'}]
--------------------
url
http://www.epicurious.com/recipes/

# Layer2
## Structure
Layer2 = List of layers [layer, layer, ..., layer]

Layer
```js
{
  id: String,   // refers to an id in layer 1
  images: [ {
    id: String, // unique 10-digit hex + .jpg
    url: String
  } ]
}
```

## * Notice Number of lists in Layer1 (1029720) and Layer2 (402760) are different
I guess this means that some layers in Layer1 do not have images.

# Ingredient

In [6]:
ingr[0]

{'valid': [True,
  True,
  True,
  True,
  True,
  True,
  True,
  True,
  False,
  False,
  True,
  True,
  True,
  False],
 'id': '000018c8a5',
 'ingredients': [{'text': 'penne'},
  {'text': 'cheese sauce'},
  {'text': 'cheddar cheese'},
  {'text': 'gruyere cheese'},
  {'text': 'dried chipotle powder'},
  {'text': 'unsalted butter'},
  {'text': 'all - purpose flour'},
  {'text': 'milk'},
  {'text': '14 ounces semihard cheese (page 23), grated (about 3 1/2 cups)'},
  {'text': '2 ounces semisoft cheese (page 23), grated (1/2 cup)'},
  {'text': 'kosher salt'},
  {'text': 'dried chipotle powder'},
  {'text': 'garlic powder'},
  {'text': '(makes about 4 cups)'}]}

In [7]:
ingr[1]

{'valid': [True, True, True, True, True, True, True, True, True],
 'id': '000033e39b',
 'ingredients': [{'text': 'elbow macaroni'},
  {'text': 'American cheese'},
  {'text': 'celery'},
  {'text': 'green peppers'},
  {'text': 'pimentos'},
  {'text': 'mayonnaise'},
  {'text': 'vinegar'},
  {'text': 'salt'},
  {'text': 'dry dill weed'}]}

# Now let's look at the Layer1 and Layer2 with the same id (==00003a70b1)

In [8]:
print('No. of Layers: %i'%(len(layer2)))
print('-'*100)
print('Example: index==0')
print('-'*100)

for k, v in layer2[0].items():
    print(k)
    print(v)
    print('-'*20)
print('-'*100)

No. of Layers: 402760
----------------------------------------------------------------------------------------------------
Example: index==0
----------------------------------------------------------------------------------------------------
id
00003a70b1
--------------------
images
[{'id': '3e233001e2.jpg', 'url': 'http://img.sndimg.com/food/image/upload/w_512,h_512,c_fit,fl_progressive,q_95/v1/img/recipes/47/91/49/picaYYmb9.jpg'}, {'id': '7f749987f9.jpg', 'url': 'http://img.sndimg.com/food/image/upload/w_512,h_512,c_fit,fl_progressive,q_95/v1/img/recipes/47/91/49/picpy37SW.jpg'}, {'id': 'aaf6b2dcd3.jpg', 'url': 'http://img.sndimg.com/food/image/upload/w_512,h_512,c_fit,fl_progressive,q_95/v1/img/recipes/47/91/49/picX9CNE2.jpg'}]
--------------------
----------------------------------------------------------------------------------------------------


In [9]:
for layer in layer1:
    if layer['id']=='00003a70b1':
        for k, v in layer.items():
            print(k)
            print(v)
            print('-'*20)

ingredients
[{'text': '2 12 cups milk'}, {'text': '1 12 cups water'}, {'text': '14 cup butter'}, {'text': 'mashed potatoes, 1 box, homestyle'}, {'text': '1 (8 ounce) can whole kernel corn (drained)'}, {'text': '1 cup cheddar cheese'}, {'text': '1 cup French-fried onions'}]
--------------------
url
http://www.food.com/recipe/crunchy-onion-potato-bake-479149
--------------------
partition
test
--------------------
title
Crunchy Onion Potato Bake
--------------------
id
00003a70b1
--------------------
instructions
[{'text': 'Preheat oven to 350 degrees Fahrenheit.'}, {'text': 'Spray pan with non stick cooking spray.'}, {'text': 'Heat milk, water and butter to boiling; stir in contents of both pouches of potatoes; let stand one minute.'}, {'text': 'Stir in corn.'}, {'text': 'Spoon half the potato mixture in pan.'}, {'text': 'Sprinkle half each of cheese and onions; top with remaining potatoes.'}, {'text': 'Sprinkle with remaining cheese and onions.'}, {'text': 'Bake 10 to 15 minutes until 

# Let's load images (for id==00003a70b1)
These are images from 1 menu (id=00003a70b1)

In [10]:
for l2 in layer2:
    if l2['id']=='00003a70b1':
        for image in l2['images']:
            print(image['id'], image['url'])
            img = Image(url=image['url']) 
            display(img)

3e233001e2.jpg http://img.sndimg.com/food/image/upload/w_512,h_512,c_fit,fl_progressive,q_95/v1/img/recipes/47/91/49/picaYYmb9.jpg


7f749987f9.jpg http://img.sndimg.com/food/image/upload/w_512,h_512,c_fit,fl_progressive,q_95/v1/img/recipes/47/91/49/picpy37SW.jpg


aaf6b2dcd3.jpg http://img.sndimg.com/food/image/upload/w_512,h_512,c_fit,fl_progressive,q_95/v1/img/recipes/47/91/49/picX9CNE2.jpg


# Dataset Stats

In [11]:
df1 = pd.read_json('/projects/kawa7978/layer1.json')
df1.rename(columns={'id':'recipe_id'}, inplace=True)

In [12]:
df2 = pd.read_json('/projects/kawa7978/layer2.json')
df2.rename(columns={'id':'recipe_id'}, inplace=True)

In [13]:
df3 = pd.read_json('/projects/kawa7978/det_ingrs.json')
df3.rename(columns={'id':'recipe_id'}, inplace=True)

In [14]:
df1.head()

ingredients  \
0  [{'text': '6 ounces penne'}, {'text': '2 cups ...   
1  [{'text': '1 c. elbow macaroni'}, {'text': '1 ...   
2  [{'text': '8 tomatoes, quartered'}, {'text': '...   
3  [{'text': '2 12 cups milk'}, {'text': '1 12 cu...   
4  [{'text': '1 (3 ounce) package watermelon gela...   

                                                 url partition  \
0  http://www.epicurious.com/recipes/food/views/-...     train   
1  http://cookeatshare.com/recipes/dilly-macaroni...     train   
2  http://www.foodnetwork.com/recipes/gazpacho1.html     train   
3  http://www.food.com/recipe/crunchy-onion-potat...      test   
4  http://www.food.com/recipe/cool-n-easy-creamy-...     train   

                                title   recipe_id  \
0          Worlds Best Mac and Cheese  000018c8a5   
1         Dilly Macaroni Salad Recipe  000033e39b   
2                            Gazpacho  000035f7ed   
3           Crunchy Onion Potato Bake  00003a70b1   
4  Cool 'n Easy Creamy Watermelon Pie  00004320bb   

                                        instructions  
0  [{'text': 'Preheat the oven to 350 F. Butter o...  
1  [{'text': 'Cook macaroni according to package ...  
2  [{'text': 'Add the tomatoes to a food processo...  
3  [{'text': 'Preheat oven to 350 degrees Fahrenh...  
4  [{'text': 'Dissolve Jello in boiling water.'},...

In [15]:
df2.head()

recipe_id                                             images
0  00003a70b1  [{'id': '3e233001e2.jpg', 'url': 'http://img.s...
1  000075604a  [{'id': '6bdca6e490.jpg', 'url': 'https://img-...
2  00007bfd16  [{'id': '6409eab844.jpg', 'url': 'http://img.s...
3  000095fc1d  [{'id': 'a1374cdd98.jpg', 'url': 'http://tasty...
4  0000b1e2b5  [{'id': 'cb1a684683.jpg', 'url': 'http://asset...

In [16]:
df3.head()

valid   recipe_id  \
0  [True, True, True, True, True, True, True, Tru...  000018c8a5   
1  [True, True, True, True, True, True, True, Tru...  000033e39b   
2  [True, True, True, True, True, True, True, Tru...  000035f7ed   
3         [True, True, True, True, True, True, True]  00003a70b1   
4                     [True, True, True, True, True]  00004320bb   

                                         ingredients  
0  [{'text': 'penne'}, {'text': 'cheese sauce'}, ...  
1  [{'text': 'elbow macaroni'}, {'text': 'America...  
2  [{'text': 'tomatoes'}, {'text': 'kosher salt'}...  
3  [{'text': 'milk'}, {'text': 'water'}, {'text':...  
4  [{'text': 'watermelon gelatin'}, {'text': 'boi...

## Train/Val/Test

In [17]:
dataset_sizes = df1.partition.value_counts()
dataset_sizes

train    720639
val      155036
test     154045
Name: partition, dtype: int64

In [18]:
dataset_sizes/dataset_sizes.sum()

train    0.699840
val      0.150561
test     0.149599
Name: partition, dtype: float64

# Now let's just retrieve 5 or 10 % of the dataset

## set number of dataset

In [19]:
percent = 0.01 # start with 1 %

num_train = dataset_sizes.train # plug above result here
num_val = dataset_sizes.val
num_test = dataset_sizes.test

exp_num_train = int(percent*num_train) # Export number of train dataset
exp_num_val = int(percent*num_val)
exp_num_test = int(percent*num_test)

print('1% Dataset size')
print(exp_num_train, exp_num_val, exp_num_test)
print('Total: %i'%(exp_num_train + exp_num_val + exp_num_test))

1% Dataset size
7206 1550 1540
Total: 10296


## get index from the original Layer1 and create new export layer 1 exp_df1

In [20]:
exp_df1 = pd.concat([
    df1[df1.partition=='train'].sample(exp_num_train),
    df1[df1.partition=='val'].sample(exp_num_val),
    df1[df1.partition=='test'].sample(exp_num_test)
])
exp_df1.head()

ingredients  \
84120   [{'text': '3 tablespoons flour'}, {'text': '1 ...   
234556  [{'text': '4 boneless skinless chicken breasts...   
94411   [{'text': '3 cups oats, toasted (I use the qui...   
746197  [{'text': '5 ounces broccoli, cut into small p...   
977611  [{'text': '1 12 cups rolled oats, not the quic...   

                                                      url partition  \
84120   http://www.food.com/recipe/old-fashioned-lemon...     train   
234556  http://www.food.com/recipe/broccoli-cheese-chi...     train   
94411   http://www.food.com/recipe/crunchy-granola-bar...     train   
746197  http://www.food.com/recipe/tortellini-pesto-an...     train   
977611  http://www.food.com/recipe/autumn-celebration-...     train   

                                       title   recipe_id  \
84120                Old-Fashioned Lemon Pie  14dd4ab991   
234556               Broccoli Cheese Chicken  3a45bbac33   
94411                   Crunchy Granola Bars  1766e23ed7   
746197  Tortellini, Pesto and Broccoli Salad  b97c71ec9f   
977611      Autumn Celebration Baked Oatmeal  f2f08ff936   

                                             instructions  
84120   [{'text': 'Preheat oven to 450F.'}, {'text': '...  
234556  [{'text': 'Begin cooking rice according to pac...  
94411   [{'text': 'To toast oats spread them on an ung...  
746197  [{'text': 'Bringa large pan of water to the bo...  
977611  [{'text': 'Preheat oven to 350 degrees butter ...

In [21]:
exp_df1.describe()

ingredients  \
count                                               10296   
unique                                              10296   
top     [{'text': '23 cup white rice'}, {'text': '4 cu...   
freq                                                    1   

                                                      url partition  \
count                                               10296     10296   
unique                                              10295         3   
top     http://www.kraftrecipes.com/recipes/taco-pizza...     train   
freq                                                    2      7206   

                         title   recipe_id  \
count                    10296       10296   
unique                   10150       10296   
top     Sweet Potato Casserole  6999a97569   
freq                         4           1   

                                             instructions  
count                                               10296  
unique                                              10288  
top     [{'text': 'Put ingredients into bread machine ...  
freq                                                    3

## From exp_df1, get exp_df2

In [22]:
# Find rows that have same IDs. 
exp_df2 = df2.merge(exp_df1, on='recipe_id')[['recipe_id', 'images']]
exp_df2.head()

recipe_id                                             images
0  0024419cd2  [{'id': '778025ea27.jpg', 'url': 'http://tasty...
1  002a3b4c85  [{'id': '8521419996.jpg', 'url': 'http://img.s...
2  00451d4108  [{'id': 'b8c271de69.jpg', 'url': 'http://tasty...
3  005b741ddb  [{'id': 'd6cd535df8.jpg', 'url': 'http://img.s...
4  006499997c  [{'id': 'c26f0ff2a1.jpg', 'url': 'http://img.s...

In [23]:
exp_df2.describe()

recipe_id                                             images
count         4092                                               4092
unique        4092                                               4092
top     0eec01ddb3  [{'id': 'a19a331f15.jpg', 'url': 'http://cdn-i...
freq             1                                                  1

In [27]:
indexes = df3.merge(exp_df1, on='recipe_id').index
exp_df3 = df3.loc[indexes]
exp_df3.head()

valid   recipe_id  \
0  [True, True, True, True, True, True, True, Tru...  000018c8a5   
1  [True, True, True, True, True, True, True, Tru...  000033e39b   
2  [True, True, True, True, True, True, True, Tru...  000035f7ed   
3         [True, True, True, True, True, True, True]  00003a70b1   
4                     [True, True, True, True, True]  00004320bb   

                                         ingredients  
0  [{'text': 'penne'}, {'text': 'cheese sauce'}, ...  
1  [{'text': 'elbow macaroni'}, {'text': 'America...  
2  [{'text': 'tomatoes'}, {'text': 'kosher salt'}...  
3  [{'text': 'milk'}, {'text': 'water'}, {'text':...  
4  [{'text': 'watermelon gelatin'}, {'text': 'boi...

## From exp_df2, select images from the dataset and create new datasets
I might need to do this in the local environment

In [28]:
exp_df1.to_json('/projects/kawa7978/small_layer1.json', orient='records')
exp_df2.to_json('/projects/kawa7978/small_layer2.json', orient='records')
exp_df3.to_json('/projects/kawa7978/small_det_ingrs.json', orient='records')